### PR TITLE
ancestor hashes socket ping/pong support

### LINKED ISSUE #26963: Feature Gate: ancestor hashes socket ping/pong support
### Description

Feature gate is needed to enable processing of ping/pong messages to enable ping/pong support for the ancestor hashes socket. See #26866

### Feature ID

AXLB87anNaUQtqBSsxkm4gvNzYY985aLtNtpJC94uWLJ

### Activation Method

Single Core Contributor

### Minimum Beta Version

_No response_

### Minimum Stable Version

v1.11.6

### Testnet Activation Epoch

_No response_

### Devnet Activation Epoch

_No response_

### Mainnet-Beta Activation Epoch

_No response_

### PR DESCRIPTION
**Pull request by jbiseda**
#### Problem

Need ping/pong support for ancestor hashes socket.

#### Summary of Changes

- Reply to ping responses sent to the ancestor hashes socket.
- Require a ping/pong cache check for ancestor hashes requests.

Fixes #26963

### CODE REVIEW

**Comment by behzadnouri** [comment]
I thought you mentioned there is no amplification with ancestor hash requests?

File: core/src/serve_repair.rs [hunk 1]
```diff
@@ -727,6 +697,60 @@ impl ServeRepair {
...
+                )
+            }
+        }
+    }
+
+    fn request_to_ping_packet(
+        request: &RepairProtocol,
+        dest_addr: &SocketAddr,
+        ping: Ping,
+    ) -> Option<Packet> {
```
Comment (behzadnouri) [inline core/src/serve_repair.rs:731]: This is a very confusing function.
If you don't need a ping for e.g. `LegacyWindowIndex` why already create a `ping: Ping` and pass it to this function?!

Comment (jbiseda) [reply core/src/serve_repair.rs]: restructured to collapse `match` statements

File: core/src/serve_repair.rs [hunk 2]
```diff
@@ -772,18 +798,22 @@ impl ServeRepair {
...
-                continue;
+            if Self::requires_ping_check(&request, root_bank, check_ping_ancestor_request_epoch) {
+                if !ContactInfo::is_valid_address(&from_addr, &socket_addr_space) {
+                    stats.err_malformed += 1;
+                    continue;
+                }
+                let (check, ping) =
+                    Self::check_ping_cache(&request, &from_addr, &identity_keypair, ping_cache);
+                if let Some(ping) = ping {
+                    if let Some(pkt) = Self::request_to_ping_packet(&request, &from_addr, ping) {
```
[the remark this answers is not in the document]
Comment (jbiseda) [reply core/src/serve_repair.rs]: if `Pakcket::from_data` returns `None`.

File: core/src/ancestor_hashes_service.rs [hunk 3]
```diff
@@ -321,55 +341,98 @@ impl AncestorHashesService {
...
+            stats.invalid_packets += 1;
+            return None;
+        };
+
+        match response {
+            AncestorHashesResponse::Hashes(ref hashes) => {
+                if packet_buf.len().saturating_sub(position) != SIZE_OF_NONCE {
+                    stats.invalid_packets += 1;
+                    return None;
+                }
```
[the remark this answers is not in the document]
Comment (jbiseda) [reply core/src/ancestor_hashes_service.rs]: done

**Review by behzadnouri — DISMISSED** [review]
lgtm

**Pull request merged**